Problem: zgossip stops czmq from compiling

```bash
src/zgossip_msg.c:442:23: error: address of array 'self->key' will always evaluate to 'true' [-Werror,-Wpointer-bool-conversion]
            if (self->key)
```

OK, I'm on it.

On Mon, May 4, 2015 at 11:08 AM, Asmod4n <notifications@github.com> wrote:

> src/zgossip_msg.c:442:23: error: address of array 'self->key' will always
> evaluate to 'true' [-Werror,-Wpointer-bool-conversion]
> if (self->key)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/zeromq/czmq/issues/1000>.
>

--- a/api/zsock.xml
+++ b/api/zsock.xml
@@ -180,6 +180,7 @@
             c = zchunk_t *
             f = zframe_t *
             h = zhashx_t *
+            U = zuuid_t *
             p = void * (sends the pointer value, only meaningful over inproc)
             m = zmsg_t * (sends all frames in the zmsg)
             z = sends zero-sized frame (0 arguments)
@@ -216,8 +217,9 @@
             b = byte **, size_t * (2 arguments) (allocates memory)
             c = zchunk_t ** (creates zchunk)
             f = zframe_t ** (creates zframe)
-            p = void ** (stores pointer)
+            U = zuuid_t * (creates a zuuid with the data)
             h = zhashx_t ** (creates zhashx)
+            p = void ** (stores pointer)
             m = zmsg_t ** (creates a zmsg with the remaing frames)
             z = null, asserts empty frame (0 arguments)
             u = uint * (stores unsigned integer, deprecated)
@@ -260,6 +262,7 @@
             S       char *, 0-2^32-1 chars  type = "longstr"
             c       zchunk_t *              type = "chunk"
             f       zframe_t *              type = "frame"
+            u       zuuid_t *               type = "uuid"
             m       zmsg_t *                type = "msg"
             p       void *, sends pointer value, only over inproc
 

--- a/bindings/python/czmq.py
+++ b/bindings/python/czmq.py
@@ -1881,6 +1881,7 @@ of these characters, each corresponding to one or two arguments:
     c = zchunk_t *
     f = zframe_t *
     h = zhashx_t *
+    U = zuuid_t *
     p = void * (sends the pointer value, only meaningful over inproc)
     m = zmsg_t * (sends all frames in the zmsg)
     z = sends zero-sized frame (0 arguments)
@@ -1911,8 +1912,9 @@ a series of pointers as provided by the caller:
     b = byte **, size_t * (2 arguments) (allocates memory)
     c = zchunk_t ** (creates zchunk)
     f = zframe_t ** (creates zframe)
-    p = void ** (stores pointer)
+    U = zuuid_t * (creates a zuuid with the data)
     h = zhashx_t ** (creates zhashx)
+    p = void ** (stores pointer)
     m = zmsg_t ** (creates a zmsg with the remaing frames)
     z = null, asserts empty frame (0 arguments)
     u = uint * (stores unsigned integer, deprecated)
@@ -1949,6 +1951,7 @@ type of each argument. Supports these argument types:
     S       char *, 0-2^32-1 chars  type = "longstr"
     c       zchunk_t *              type = "chunk"
     f       zframe_t *              type = "frame"
+    u       zuuid_t *               type = "uuid"
     m       zmsg_t *                type = "msg"
     p       void *, sends pointer value, only over inproc
 

--- a/bindings/qml/src/QmlZsock.cpp
+++ b/bindings/qml/src/QmlZsock.cpp
@@ -95,6 +95,7 @@ const QString QmlZsock::typeStr () {
 //      c = zchunk_t *                                                    
 //      f = zframe_t *                                                    
 //      h = zhashx_t *                                                    
+//      U = zuuid_t *                                                     
 //      p = void * (sends the pointer value, only meaningful over inproc) 
 //      m = zmsg_t * (sends all frames in the zmsg)                       
 //      z = sends zero-sized frame (0 arguments)                          
@@ -129,8 +130,9 @@ int QmlZsock::vsend (const QString &picture, va_list argptr) {
 //      b = byte **, size_t * (2 arguments) (allocates memory)              
 //      c = zchunk_t ** (creates zchunk)                                    
 //      f = zframe_t ** (creates zframe)                                    
-//      p = void ** (stores pointer)                                        
+//      U = zuuid_t * (creates a zuuid with the data)                       
 //      h = zhashx_t ** (creates zhashx)                                    
+//      p = void ** (stores pointer)                                        
 //      m = zmsg_t ** (creates a zmsg with the remaing frames)              
 //      z = null, asserts empty frame (0 arguments)                         
 //      u = uint * (stores unsigned integer, deprecated)                    
@@ -171,6 +173,7 @@ int QmlZsock::vrecv (const QString &picture, va_list argptr) {
 //      S       char *, 0-2^32-1 chars  type = "longstr"                   
 //      c       zchunk_t *              type = "chunk"                     
 //      f       zframe_t *              type = "frame"                     
+//      u       zuuid_t *               type = "uuid"                      
 //      m       zmsg_t *                type = "msg"                       
 //      p       void *, sends pointer value, only over inproc              
 //                                                                         

--- a/bindings/qml/src/QmlZsock.h
+++ b/bindings/qml/src/QmlZsock.h
@@ -93,6 +93,7 @@ public slots:
     //      c = zchunk_t *                                                    
     //      f = zframe_t *                                                    
     //      h = zhashx_t *                                                    
+    //      U = zuuid_t *                                                     
     //      p = void * (sends the pointer value, only meaningful over inproc) 
     //      m = zmsg_t * (sends all frames in the zmsg)                       
     //      z = sends zero-sized frame (0 arguments)                          
@@ -121,8 +122,9 @@ public slots:
     //      b = byte **, size_t * (2 arguments) (allocates memory)              
     //      c = zchunk_t ** (creates zchunk)                                    
     //      f = zframe_t ** (creates zframe)                                    
-    //      p = void ** (stores pointer)                                        
+    //      U = zuuid_t * (creates a zuuid with the data)                       
     //      h = zhashx_t ** (creates zhashx)                                    
+    //      p = void ** (stores pointer)                                        
     //      m = zmsg_t ** (creates a zmsg with the remaing frames)              
     //      z = null, asserts empty frame (0 arguments)                         
     //      u = uint * (stores unsigned integer, deprecated)                    
@@ -157,6 +159,7 @@ public slots:
     //      S       char *, 0-2^32-1 chars  type = "longstr"                   
     //      c       zchunk_t *              type = "chunk"                     
     //      f       zframe_t *              type = "frame"                     
+    //      u       zuuid_t *               type = "uuid"                      
     //      m       zmsg_t *                type = "msg"                       
     //      p       void *, sends pointer value, only over inproc              
     //                                                                         

--- a/bindings/ruby/lib/czmq/ffi/zsock.rb
+++ b/bindings/ruby/lib/czmq/ffi/zsock.rb
@@ -268,6 +268,7 @@ module CZMQ
       #     c = zchunk_t *                                                    
       #     f = zframe_t *                                                    
       #     h = zhashx_t *                                                    
+      #     U = zuuid_t *                                                     
       #     p = void * (sends the pointer value, only meaningful over inproc) 
       #     m = zmsg_t * (sends all frames in the zmsg)                       
       #     z = sends zero-sized frame (0 arguments)                          
@@ -306,8 +307,9 @@ module CZMQ
       #     b = byte **, size_t * (2 arguments) (allocates memory)              
       #     c = zchunk_t ** (creates zchunk)                                    
       #     f = zframe_t ** (creates zframe)                                    
-      #     p = void ** (stores pointer)                                        
+      #     U = zuuid_t * (creates a zuuid with the data)                       
       #     h = zhashx_t ** (creates zhashx)                                    
+      #     p = void ** (stores pointer)                                        
       #     m = zmsg_t ** (creates a zmsg with the remaing frames)              
       #     z = null, asserts empty frame (0 arguments)                         
       #     u = uint * (stores unsigned integer, deprecated)                    
@@ -352,6 +354,7 @@ module CZMQ
       #     S       char *, 0-2^32-1 chars  type = "longstr"                   
       #     c       zchunk_t *              type = "chunk"                     
       #     f       zframe_t *              type = "frame"                     
+      #     u       zuuid_t *               type = "uuid"                      
       #     m       zmsg_t *                type = "msg"                       
       #     p       void *, sends pointer value, only over inproc              
       #                                                                        

--- a/include/zframe.h
+++ b/include/zframe.h
@@ -35,10 +35,10 @@ CZMQ_EXPORT zframe_t *
 CZMQ_EXPORT void
     zframe_destroy (zframe_t **self_p);
 
-//  Create an empty (zero-sized) frame
-//  The caller is responsible for destroying the return value when finished with it.
+//  Create an empty (zero-sized) frame. The caller is responsible for 
+//  destroying the return value when finished with it.
 CZMQ_EXPORT zframe_t *
-    zframe_new_empty ();
+    zframe_new_empty (void);
 
 //  Receive frame from socket, returns zframe_t object or NULL if the recv  
 //  was interrupted. Does a blocking recv, if you want to not block then use

--- a/include/zsock.h
+++ b/include/zsock.h
@@ -174,6 +174,7 @@ CZMQ_EXPORT const char *
 //      c = zchunk_t *                                                    
 //      f = zframe_t *                                                    
 //      h = zhashx_t *                                                    
+//      U = zuuid_t *                                                     
 //      p = void * (sends the pointer value, only meaningful over inproc) 
 //      m = zmsg_t * (sends all frames in the zmsg)                       
 //      z = sends zero-sized frame (0 arguments)                          
@@ -204,8 +205,9 @@ CZMQ_EXPORT int
 //      b = byte **, size_t * (2 arguments) (allocates memory)              
 //      c = zchunk_t ** (creates zchunk)                                    
 //      f = zframe_t ** (creates zframe)                                    
-//      p = void ** (stores pointer)                                        
+//      U = zuuid_t * (creates a zuuid with the data)                       
 //      h = zhashx_t ** (creates zhashx)                                    
+//      p = void ** (stores pointer)                                        
 //      m = zmsg_t ** (creates a zmsg with the remaing frames)              
 //      z = null, asserts empty frame (0 arguments)                         
 //      u = uint * (stores unsigned integer, deprecated)                    
@@ -242,6 +244,7 @@ CZMQ_EXPORT int
 //      S       char *, 0-2^32-1 chars  type = "longstr"                   
 //      c       zchunk_t *              type = "chunk"                     
 //      f       zframe_t *              type = "frame"                     
+//      u       zuuid_t *               type = "uuid"                      
 //      m       zmsg_t *                type = "msg"                       
 //      p       void *, sends pointer value, only over inproc              
 //                                                                         
@@ -307,7 +310,7 @@ CZMQ_EXPORT void
 
 // zsock leak detection - not a part of the official interface to zsock. This
 // enables CZMQ to report socket leaks intelligently.
-#if defined ZSOCK_NOCHECK 
+#if defined ZSOCK_NOCHECK
     // no checking active - use the above interface methods directly.
 #else
 #   define zsock_new(t) zsock_new_checked((t), __FILE__, __LINE__)

--- a/src/zframe.c
+++ b/src/zframe.c
@@ -61,7 +61,8 @@ zframe_new (const void *data, size_t size)
 
 
 //  --------------------------------------------------------------------------
-//  Constructor; Allocates a new empty (zero-sized) frame
+//  Create an empty (zero-sized) frame. The caller is responsible for 
+//  destroying the return value when finished with it.
 
 zframe_t *
 zframe_new_empty (void)

--- a/src/zgossip_engine.inc
+++ b/src/zgossip_engine.inc
@@ -738,7 +738,7 @@ s_server_config_service (s_server_t *self)
             if (zsock_bind (self->router, "%s", endpoint) == -1)
                 zsys_warning ("could not bind to %s (%s)", endpoint, zmq_strerror (zmq_errno ()));
         }
-#if (ZMQ_MAJOR_VERSION >= 4)
+#if (ZMQ_VERSION_MAJOR >= 4)
         else
         if (streq (zconfig_name (section), "security")) {
             char *mechanism = zconfig_resolve (section, "mechanism", "null");

--- a/src/zgossip_msg.c
+++ b/src/zgossip_msg.c
@@ -39,12 +39,9 @@ struct _zgossip_msg_t {
     int id;                             //  zgossip_msg message ID
     byte *needle;                       //  Read/write pointer for serialization
     byte *ceiling;                      //  Valid upper limit for read pointer
-    /* Tuple key, globally unique  */
-    char key [256];
-    /* Tuple value, as printable string  */
-    char *value;
-    /* Time to live, msecs  */
-    uint32_t ttl;
+    char key [256];                     //  Tuple key, globally unique
+    char *value;                        //  Tuple value, as printable string
+    uint32_t ttl;                       //  Time to live, msecs
 };
 
 //  --------------------------------------------------------------------------
@@ -237,7 +234,7 @@ int
 zgossip_msg_recv (zgossip_msg_t *self, zsock_t *input)
 {
     assert (input);
-    
+
     if (zsock_type (input) == ZMQ_ROUTER) {
         zframe_destroy (&self->routing_id);
         self->routing_id = zframe_recv (input);
@@ -256,7 +253,7 @@ zgossip_msg_recv (zgossip_msg_t *self, zsock_t *input)
     //  Get and check protocol signature
     self->needle = (byte *) zmq_msg_data (&frame);
     self->ceiling = self->needle + zmq_msg_size (&frame);
-    
+
     uint16_t signature;
     GET_NUMBER2 (signature);
     if (signature != (0xAAA0 | 0)) {
@@ -386,7 +383,7 @@ zgossip_msg_send (zgossip_msg_t *self, zsock_t *output)
     PUT_NUMBER2 (0xAAA0 | 0);
     PUT_NUMBER1 (self->id);
     size_t nbr_frames = 1;              //  Total number of frames to send
-    
+
     switch (self->id) {
         case ZGOSSIP_MSG_HELLO:
             PUT_NUMBER1 (1);
@@ -418,7 +415,7 @@ zgossip_msg_send (zgossip_msg_t *self, zsock_t *output)
     }
     //  Now send the data frame
     zmq_msg_send (&frame, zsock_resolve (output), --nbr_frames? ZMQ_SNDMORE: 0);
-    
+
     return 0;
 }
 
@@ -435,36 +432,33 @@ zgossip_msg_print (zgossip_msg_t *self)
             zsys_debug ("ZGOSSIP_MSG_HELLO:");
             zsys_debug ("    version=1");
             break;
-            
+
         case ZGOSSIP_MSG_PUBLISH:
             zsys_debug ("ZGOSSIP_MSG_PUBLISH:");
             zsys_debug ("    version=1");
-            if (self->key)
-                zsys_debug ("    key='%s'", self->key);
-            else
-                zsys_debug ("    key=");
+            zsys_debug ("    key='%s'", self->key);
             if (self->value)
                 zsys_debug ("    value='%s'", self->value);
             else
                 zsys_debug ("    value=");
             zsys_debug ("    ttl=%ld", (long) self->ttl);
             break;
-            
+
         case ZGOSSIP_MSG_PING:
             zsys_debug ("ZGOSSIP_MSG_PING:");
             zsys_debug ("    version=1");
             break;
-            
+
         case ZGOSSIP_MSG_PONG:
             zsys_debug ("ZGOSSIP_MSG_PONG:");
             zsys_debug ("    version=1");
             break;
-            
+
         case ZGOSSIP_MSG_INVALID:
             zsys_debug ("ZGOSSIP_MSG_INVALID:");
             zsys_debug ("    version=1");
             break;
-            
+
     }
 }
 
@@ -598,7 +592,7 @@ zgossip_msg_set_ttl (zgossip_msg_t *self, uint32_t ttl)
 int
 zgossip_msg_test (bool verbose)
 {
-    printf (" * zgossip_msg: ");
+    printf (" * zgossip_msg:");
 
     //  Silence an "unused" warning by "using" the verbose variable
     if (verbose) {;}

--- a/src/zgossip_msg.h
+++ b/src/zgossip_msg.h
@@ -1,6 +1,6 @@
 /*  =========================================================================
     zgossip_msg - ZeroMQ Gossip Protocol
-    
+
     Codec header for zgossip_msg.
 
     ** WARNING *************************************************************
@@ -84,7 +84,7 @@ CZMQ_EXPORT int
 //  Send the zgossip_msg to the output socket, does not destroy it
 CZMQ_EXPORT int
     zgossip_msg_send (zgossip_msg_t *self, zsock_t *output);
-    
+
 //  Print contents of message to stdout
 CZMQ_EXPORT void
     zgossip_msg_print (zgossip_msg_t *self);

--- a/src/zpoller.c
+++ b/src/zpoller.c
@@ -25,7 +25,7 @@
 //  Structure of our class
 
 struct _zpoller_t {
-    zlist_t *reader_list;      //  List of sockets to read from
+    zlist_t *reader_list;       //  List of sockets to read from
     zmq_pollitem_t *poll_set;   //  Current zmq_poll set
     void **poll_readers;        //  Matching table of socket readers
     size_t poll_size;           //  Size of poll set

--- a/src/zrex.c
+++ b/src/zrex.c
@@ -259,6 +259,7 @@ zrex_test (bool verbose)
 {
     printf (" * zrex: ");
 
+    //  @selftest
     //  This shows the pattern of matching many lines to a single pattern
     zrex_t *rex = zrex_new ("\\d+-\\d+-\\d+");
     assert (rex);
@@ -301,5 +302,7 @@ zrex_test (bool verbose)
     assert (streq (zrex_hit (rex, 1), "CURVE"));
     assert (streq (mechanism, "CURVE"));
     zrex_destroy (&rex);
+    
+    //  @end
     printf ("OK\n");
 }

--- a/src/zsock.c
+++ b/src/zsock.c
@@ -625,6 +625,7 @@ zsock_type_str (zsock_t *self)
 //      c = zchunk_t *
 //      f = zframe_t *
 //      h = zhashx_t *
+//      U = zuuid_t *
 //      p = void * (sends the pointer value, only meaningful over inproc)
 //      m = zmsg_t * (sends all frames in the zmsg)
 //      z = sends zero-sized frame (0 arguments)
@@ -700,6 +701,11 @@ zsock_vsend (void *self, const char *picture, va_list argptr)
             zmsg_addmem (msg, zframe_data (frame), zframe_size (frame));
         }
         else
+        if (*picture == 'U') {
+            zuuid_t *uuid = va_arg (argptr, zuuid_t *);
+            zmsg_addmem (msg, zuuid_data (uuid), zuuid_size (uuid));
+        }
+        else
         if (*picture == 'p') {
             void *pointer = va_arg (argptr, void *);
             zmsg_addmem (msg, &pointer, sizeof (void *));
@@ -747,8 +753,9 @@ zsock_vsend (void *self, const char *picture, va_list argptr)
 //      b = byte **, size_t * (2 arguments) (allocates memory)
 //      c = zchunk_t ** (creates zchunk)
 //      f = zframe_t ** (creates zframe)
-//      p = void ** (stores pointer)
 //      h = zhashx_t ** (creates zhashx)
+//      U = zuuid_t * (creates a zuuid with the data)
+//      p = void ** (stores pointer)
 //      m = zmsg_t ** (creates a zmsg with the remaing frames)
 //      z = null, asserts empty frame (0 arguments)
 //      u = uint * (stores unsigned integer, deprecated)
@@ -888,6 +895,20 @@ zsock_vrecv (void *self, const char *picture, va_list argptr)
                 *frame_p = frame;
             else
                 zframe_destroy (&frame);
+        }
+        else
+        if (*picture == 'U') {
+            zframe_t *frame = zmsg_pop (msg);
+            zuuid_t **uuid_p = va_arg (argptr, zuuid_t **);
+            if (uuid_p) {
+                if (frame) {
+                    *uuid_p = zuuid_new ();
+                    zuuid_set (*uuid_p, zframe_data (frame));
+                }
+                else
+                    *uuid_p = NULL;
+            }
+            zframe_destroy (&frame);
         }
         else
         if (*picture == 'p') {
@@ -1043,6 +1064,7 @@ zsock_vrecv (void *self, const char *picture, va_list argptr)
 //      S       char *, 0-2^32-1 chars  type = "longstr"
 //      c       zchunk_t *              type = "chunk"
 //      f       zframe_t *              type = "frame"
+//      u       zuuid_t *               type = "uuid"
 //      m       zmsg_t *                type = "msg"
 //      p       void *, sends pointer value, only over inproc
 //
@@ -1107,6 +1129,11 @@ zsock_bsend (void *self, const char *picture, ...)
             zframe_t *frame = va_arg (argptr, zframe_t *);
             assert (nbr_frames < ZSOCK_BSEND_MAX_FRAMES - 1);
             frames [nbr_frames++] = frame;
+        }
+        else
+        if (*picptr == 'u') {
+            zuuid_t *uuid = va_arg (argptr, zuuid_t *);
+            frame_size += zuuid_size (uuid);
         }
         else
         if (*picptr == 'm') {
@@ -1194,6 +1221,14 @@ zsock_bsend (void *self, const char *picture, ...)
                 PUT_NUMBER4 (zchunk_size (chunk));
                 memcpy (needle, zchunk_data (chunk), zchunk_size (chunk));
                 needle += zchunk_size (chunk);
+            }
+        }
+        else
+        if (*picptr == 'u') {
+            zuuid_t *uuid = va_arg (argptr, zuuid_t *);
+            if (uuid) {
+                memcpy (needle, zuuid_data (uuid), zuuid_size (uuid));
+                needle += zuuid_size (uuid);
             }
         }
         picptr++;
@@ -1329,6 +1364,16 @@ zsock_brecv (void *selfish, const char *picture, ...)
                 goto malformed;
             *chunk_p = zchunk_new (needle, chunk_size);
             needle += chunk_size;
+        }
+        else
+        if (*picptr == 'u') {
+            zuuid_t **uuid_p = va_arg (argptr, zuuid_t **);
+            *uuid_p = zuuid_new ();
+            if (zuuid_size (*uuid_p) > MAX_ALLOC_SIZE
+            ||  needle + zuuid_size (*uuid_p) > (ceiling))
+                goto malformed;
+            zuuid_set (*uuid_p, needle);
+            needle += zuuid_size (*uuid_p);
         }
         else
         if (*picptr == 'f') {
@@ -1599,27 +1644,32 @@ zsock_test (bool verbose)
     assert (frame);
     zhashx_t *hash = zhashx_new ();
     assert (hash);
+    zuuid_t *uuid = zuuid_new ();
+    assert (uuid);
     zhashx_autofree (hash);
     zhashx_insert (hash, "1", "value A");
     zhashx_insert (hash, "2", "value B");
     char *original = "pointer";
 
     //  Test zsock_recv into each supported type
-    zsock_send (writer, "i1248zsbcfhp",
+    zsock_send (writer, "i1248zsbcfuhp",
                 -12345, number1, number2, number4, number8,
-                "This is a string", "ABCDE", 5, chunk, frame, hash, original);
-    zframe_destroy (&frame);
+                "This is a string", "ABCDE", 5,
+                chunk, frame, uuid, hash, original);
+    char *uuid_str = strdup (zuuid_str (uuid));
     zchunk_destroy (&chunk);
+    zframe_destroy (&frame);
+    zuuid_destroy (&uuid);
     zhashx_destroy (&hash);
-    
+
     int integer;
     byte *data;
     size_t size;
     char *pointer;
     number8 = number4 = number2 = number1 = 0;
-    rc = zsock_recv (reader, "i1248zsbcfhp",
+    rc = zsock_recv (reader, "i1248zsbcfuhp",
                      &integer, &number1, &number2, &number4, &number8,
-                     &string, &data, &size, &chunk, &frame, &hash, &pointer);
+                     &string, &data, &size, &chunk, &frame, &uuid, &hash, &pointer);
     assert (rc == 0);
     assert (integer == -12345);
     assert (number1 == 123);
@@ -1631,6 +1681,7 @@ zsock_test (bool verbose)
     assert (size == 5);
     assert (memcmp (zchunk_data (chunk), "HELLO", 5) == 0);
     assert (zchunk_size (chunk) == 5);
+    assert (streq (uuid_str, zuuid_str (uuid)));
     assert (memcmp (zframe_data (frame), "WORLD", 5) == 0);
     assert (zframe_size (frame) == 5);
     char *value = (char *) zhashx_lookup (hash, "1");
@@ -1640,6 +1691,7 @@ zsock_test (bool verbose)
     assert (original == pointer);
     free (string);
     free (data);
+    free (uuid_str);
     zframe_destroy (&frame);
     zchunk_destroy (&chunk);
     zhashx_destroy (&hash);

